### PR TITLE
BUG: heatmap now displays all classes properly, resizes dynamically

### LINKED ISSUE #143: BUG: confusion matrix class label interval breaks with large sample size
**Bug Description**
Confusion matrix labels get misaligned with very large number of class labels.

This does NOT impact results (e.g., accuracy), only the labels — so plots need to be relabeled, e.g., for publication.

The issue seems to be that the tick intervals are being automatically set, but the tick labels are not being assigned the same interval: https://github.com/qiime2/q2-sample-classifier/blob/master/q2_sample_classifier/visuals.py#L129

**Steps to reproduce the behavior**
Make a confusion matrix with a large number of classes (example below has ~30)

**Expected behavior**
Label interval should match tick interval.

**Screenshots**
image

### PR DESCRIPTION
**Pull request by ChrisKeefe**
closes #143 , supersedes https://github.com/qiime2/q2-sample-classifier/pull/180
Thanks, @danteese for helping kick this off! Hopefully, this PR finishes what you started. :slightly_smiling_face: 
Thanks @Oddant1 for wallowing in the matplotlib docs with me.

These changes allow `heatmap` to render a heatmap of an arbitrary number of classes, displaying all classes as labels, preserving a 1:1 aspect for heatmap cells, and resizing the plot dynamically to accommodate. 

---

**Questions:** 
1. Cell size - Is this a good size for the boxes/cells that make up the heatmap itself? Do they need upsizing or downsizing? Screencaps attached below, at tiny, medium, and very large.
2. Borders - We don't outline many of our plots. I assume we want to stick with that here? I've mocked an outlined one as well, in case you like. LMK
3. Proportion bar readability - take a look at the color bar in the _tiny_ size. Is that readable enough? Do we want to make it taller for tiny-size plots? (In larger plots, it scales 1:1 with the heatmap itself. )
4. Code organization - The method modified here persists in building this plot, and _not_ labeling it. Labelling happens in the calling function, `_plot_confusion_matrix`. The references to the heatmap and colorbar axes created in this diff are not passed back to the calling function, so colorbar and its attributes remain un-editable there. I don't feel this is a major loss, but the separation of the two parts of the process feels a little weird to me. Handling both the plot build and labelling in one function would mitigate this, but would require us to pass more data around. I'm inclined to keep things as is, unless someone feels re-organization would bring us maintainability or stability in the future. What do you think?

---

**Samples:** 
Tiny: 
[hm_tiny.pdf](https://github.com/qiime2/q2-sample-classifier/files/3881968/hm_tiny.pdf)

Big: 
[hm_big.pdf](https://github.com/qiime2/q2-sample-classifier/files/3881969/hm_big.pdf)

Huge:
[hm_huge.pdf](https://github.com/qiime2/q2-sample-classifier/files/3881970/hm_huge.pdf)

Bordered Big:
[hm_bordered.pdf](https://github.com/qiime2/q2-sample-classifier/files/3881971/hm_bordered.pdf)

### CODE REVIEW

**Comment by thermokarst** [comment]
Hey @ChrisKeefe - can you provide the test data used to generate your mockup PDFs? The PDFs are really helpful (thanks!), but they don't tell the whole story --- I want to lay my :eyes: on the actual QZVs before answering your questions above. Thanks!

**Comment by ChrisKeefe** [comment]
QZVs for you, @thermokarst: 
[Tiny](https://forum.qiime2.org/uploads/short-url/wgzvfk3GS4gbfUokUVT1GrjE3YC.qzv)
[Big](https://forum.qiime2.org/uploads/short-url/ic5z5qlSJn2Mn7koWjmxp5CIbs9.qzv)
[Huge](https://forum.qiime2.org/uploads/short-url/jJAfqgndBdsxaDvRM0GuVEXKXGE.qzv)

**Comment by thermokarst** [comment]
Thanks @ChrisKeefe - can I get the test data, too?

**Comment by ChrisKeefe** [comment]
> Thanks @ChrisKeefe - can I get the test data, too?
Plots generated with this, @thermokarst . 
``` python
import qiime2
from qiime2.plugins import sample_classifier as sc
import pandas as pd
from sklearn.metrics import confusion_matrix

# mdix = [i for i in 'ab']
# mdix = [i for i in 'abcde']
# mdix = [i for i in 'abcdefgh']
# mdix = [i for i in 'abcdefghijklmnopqrstuvwxyz']
mdix = [i for i in 'abcdefghijklmnopqrstuvwxyz0123456789!@$%^&*()_+-=\][{}|;:"<>']
md = pd.Series(mdix, index=pd.Index(mdix, name='sample-id'), name='class')
truth = qiime2.CategoricalMetadataColumn(md)
predictions = qiime2.Artifact.import_data('SampleData[ClassifierPredictions]', md)
cm = sc.actions.confusion_matrix(predictions, truth)
cm.visualization.save('test-matrix.qzv')
```

**Comment by thermokarst** [comment]
Dang @ChrisKeefe, you're making me work for this one! Okay, for anyone else following along at home (ahem, @nbokulich), a few test data files can be found here:

[q2-sample-classifier-test-data.zip](https://github.com/qiime2/q2-sample-classifier/files/3889252/q2-sample-classifier-test-data.zip)

File: q2_sample_classifier/visuals.py [hunk 1]
```diff
@@ -102,8 +102,27 @@ def _linear_regress(actual, pred):
 def _plot_heatmap_from_confusion_matrix(cm, palette, vmin=None, vmax=None):
     palette = _custom_palettes()[palette]
     plt.figure()
-    return sns.heatmap(cm, vmin=vmin, vmax=vmax, cmap=palette,
-                       cbar_kws={'label': 'Proportion'})
+    scaletron, labelsize, dpi = 20, 8, 100
```
Comment (thermokarst) [inline q2_sample_classifier/visuals.py]: As much as I love seeing "scaletron" here, that probably should've been removed in code when it originally was added to q2-feature-table.

```suggestion
    scaler, labelsize, dpi = 20, 8, 100
```

File: q2_sample_classifier/visuals.py [hunk 2]
```diff
@@ -102,8 +102,27 @@ def _linear_regress(actual, pred):
...
+    sns.set(rc={'xtick.labelsize': labelsize, 'ytick.labelsize': labelsize,
+            'figure.dpi': dpi})
+    fig, (ax, cax) = plt.subplots(ncols=2, constrained_layout=True)
+    heatmap = sns.heatmap(cm, vmin=vmin, vmax=vmax, cmap=palette, ax=ax,
+                          cbar_ax=cax, cbar_kws={'label': 'Proportion'},
+                          square=True, xticklabels=True, yticklabels=True)
+
+    # Resize the plot dynamically based on number of classes
+    hm_pos = ax.get_position()
+    scale = len(cm) / scaletron
```
Comment (thermokarst) [inline q2_sample_classifier/visuals.py]:
```suggestion
    scale = len(cm) / scaler
```

File: q2_sample_classifier/visuals.py [hunk 3]
```diff
@@ -102,8 +102,27 @@ def _linear_regress(actual, pred):
...
+    # set minimum color bar height for readability
+    cbar_height = scale if (scale >= .15) else .15
```
Comment (thermokarst) [inline q2_sample_classifier/visuals.py]: Please factor out the 0.15 as a local constant, and add to line 105 above (the scaler, labelsize, dpi line).

Comment (thermokarst) [reply q2_sample_classifier/visuals.py]:
```suggestion
    cbar_height = scale if (scale > .15) else .15
```

Dropping the `=` has the same effect, and is one less thing to mentally parse when reading.

Comment (ebolyen) [reply q2_sample_classifier/visuals.py]: `max(scale, 0.15)` might be clearer.

**Comment by thermokarst** [comment]
LGTM, thanks @ChrisKeefe!
